sstables/index_reader.hh

```diff
@@ -480,6 +480,7 @@ class index_reader {
     }
 
     future<> advance_context(index_bound& bound, uint64_t begin, uint64_t end, int quantity) {
+        assert(!bound.context || !_single_page_read);
         if (!bound.context) {
             bound.consumer = std::make_unique<index_consumer>(_region, _sstable->get_schema());
```

Comment:
> But there is no use for this. The idea of a "single-partition reader"
> exist as an optimization. It's illegal to forward single-partition readers,
> and it doesn't make sense to attempt that. (If there's a need for forwarding,
> just don't create a single-partition reader).

The problem is that currently sstable readers ignore the forwarding option and always create single-partition readers if the range is singular, even if forwarding is requested:

See in sstables/mx/reader.cc
```
            // FIXME: I want to add `&& fwd_mr == mutation_reader::forwarding::no` below
            // but can't because many call sites use the default value for
            // `mutation_reader::forwarding` which is `yes`.
            , _single_partition_read(pr.is_singular())
``` 

Otherwise, all sstable reads which go through cache would be unoptimized.

So I would expect this change to break forwarding if the first range was singular.  The API of mutation_source doesn't prohibit this. A safer fix would be to in addition to this, reset the _single_page_read flag when sstable reader is fast-forwarded in `advance_to(const dht::partition_range& range)`.

We should also compare before/after for perf_fast_forward tests, and harden this workload by adding new test cases.

Reply:
> So I would expect this change to break forwarding if the first range was singular. The API of mutation_source doesn't prohibit this.

But I thought that forwarding a single-partition reader is already supposed to be illegal. If it's illegal, then the patch can't break it.

See for example the comment in: https://github.com/scylladb/scylladb/blob/5a470b2bfbe6fdb371bbde2a2f8987d6618dd265/readers/mutation_reader.hh#L343.

Or see https://github.com/scylladb/scylladb/blob/5a470b2bfbe6fdb371bbde2a2f8987d6618dd265/row_cache.cc#L395-L397, where the `row_cache` explicitly chooses to pass a singular [pk] or a non-singular [pk, pk] to the sstable reader, depending on whether it needs a forwardable reader or not.

Reply:
This workaround was added for the reason @tgrabiec mentioned. It looks like in many places there's indeed an assumption that if we create a reader with a singular partition range it cannot be forwarded, but it wasn't always the case, as the comment in sstables/mx/reader.cc shows and I'm not sure it's true everywhere now - for example in the `flat_multi_range_mutation_reader` I see no consideration about making sure the first range is changed to a non-singular one if we use it with multiple ranges. Unless we investigate all uses of the sstable reader (which is a lot) I don't think we can assume that single-partition reads are really non-forwardable, especially considering the conflicting comments. And after we do that we should remove the old comment

Also, the fix for forwardable readers that start with a singular range should probably rely on setting the right `mutation_reader::forwarding` instead of the default `yes` when creating the reader. Otherwise we risk getting a singular first range by chance somewhere and facing the same issue again

Reply:
> > but it wasn't always the case, as the comment in sstables/mx/reader.cc shows
> > especially considering the conflicting comments
> 
> Which comments are you talking about?

```
            // FIXME: I want to add `&& fwd_mr == mutation_reader::forwarding::no` below
            // but can't because many call sites use the default value for
            // `mutation_reader::forwarding` which is `yes`.
            , _single_partition_read(pr.is_singular())
```
I suppose it's only saying that we may have `mutation_reader::forwarding::yes` even when we want to do a single partition read, but I assume it was added because we end up not checking `fwd_mr` at all and can set `_single_partition_read` even if it actually forwarding

Reply:
> But I thought that forwarding a single-partition reader is already supposed to be illegal. If it's illegal, then the patch can't break it.

Yes, we could lawyer away this concern based on the comment on mutation_reader. However, I think long term we should move away from this contract as it is awkward and risky. It's too easy to break this contract for a mutation_source user, and it's a widely exposed API. We will sleep better if index_reader doesn't rely on this assumption, and resets the flag when it's attempted to be fast-forwarded, which seems like an easy thing to do. Or at least throw an exception on fast forwarding attempt in this case. I don't mean the attempt like the old code did in advance_to_page(), but the path which is called only from fast_forward(const partition_range&).